### ISSUE TITLE
Transactions support

### ISSUE THREAD

**Issue by nappa85** [SeaQL/sea-orm#142]
Coming from https://github.com/SeaQL/sea-orm/pull/118, I start a new issue con go on with the analisys:

What @tyt2y3 proposed:
> I think transaction is highly driver specific. I don't think we can cleanly wrap and generalizes over different database+driver.
> Perhaps we can start by exposing some of the underlying types and methods from SQLx.
> 
> https://docs.rs/sqlx/0.5.7/sqlx/trait.Connection.html#method.transaction

It can be a starting point, if we're able to maintain sea-orm typing.
What I mean is, sqlx Connection::transaction takes an fn(&mut Transaction), we would need to wrap it up so inside the function we can still use our DatabaseConnection type.

And it's kind of an overkill IMHO, Transaction is only a sugar wrapper over a normal connection, where a START TRANSACTION has been executed.

I would prefere an approach like this: https://docs.rs/mysql_async/0.28.0/mysql_async/struct.Conn.html#method.start_transaction
where Transaction is simply a tiny wrapper over Connection and can be used in any method that takes `&Conn` because it `impl Deref<Conn>`

About isolation levels, mysql and postgres are quite similars:
https://dev.mysql.com/doc/refman/8.0/en/set-transaction.html
https://www.postgresql.org/docs/9.5/sql-set-transaction.html

Sqlite has transactions too, it has a little different dialect, like BEGIN [TRANSACTION] instead of START TRANSACTION, and isolation levels have to be set using the PRAGMA keyword, like https://www.sqlite.org/pragma.html#pragma_read_uncommitted

**Comment by tyt2y3** [comment]
There is a reason that SQLx takes in a `FnOnce` for transaction. It is using a connection pool, this is to prevent that the user forget to COMMIT or ROLLBACK thereafter.

I think, probably, the SeaORM API can take a [`Vec<Statement>`](https://docs.rs/sea-orm/0.x/sea_orm/struct.Statement.html) and execute them in a single transaction. This is rather simplistic, and does not cover all use cases.

All in all, I think transaction in an async context is a complex matters that requires more investigation.

**Comment by nappa85** [comment]
Transactions are often used to lock to a certain state (isolation) to be sure that there are not changes between select queries. How do you cover that case with a `Vec<Statement>`?

I can try to see if I can fit inside sqlx apis with a wrapper, if it covers all scenarios

**Comment by tyt2y3** [comment]
You are right. I think it is still useful when there are multiple inserts and want to maintain 'domain consistency'.
But definitely, `FnOnce` is the most flexible way.

**Comment by nappa85** [comment]
Just to inform you, I've found a lifetime problem with my transaction implementation.
When capturing references from outside you'll end up in an impossible lifetime.
I've added a test that replicates the situation and I'm trying to solve it.
I'll post a patch as soon as I find a solution

**Comment by Stumblinbear** [comment]
Is it possible to allow returning custom error types from the transaction function? There are cases where you'd want to know how and when a transaction failed

**Comment by nappa85** [comment]
My implementation allows you to return anything that impl Error, wrapper in an error Type that permits you to know if the error happened opening the transaction

**Comment by Stumblinbear** [comment]
> My implementation allows you to return anything that impl Error, wrapper in an error Type that permits you to know if the error happened opening the transaction

Yeah, just noticed that, haha

**Comment by Stumblinbear** [comment]
I managed to work around the lifetime issue by cloning the Arc just before starting the transaction, so thankfully that's a non-issue for me. I do have an additional question, though. Is it possible to accept both a normal connection and a transaction connection in a single function?

**Comment by nappa85** [comment]
Yes, use a generic over trait ConnectionTrait

**Comment by Stumblinbear** [comment]
Hm. Well, I guess my inexperience with Rust is the issue, here. Anyway, thanks for the tip. Seems like you thought of everything, already, haha

**Comment by Stumblinbear** [comment]
I've got a suggestion, if it's in scope for this change. Could there be support for running tests inside of a transaction that always rolls back once the function has completed? Perhaps a procedural macro? The current way to do this is to start a transaction in each test, and always return Rollback in an Err, but this is a bit awkward

**Comment by tyt2y3** [comment]
The theme of [0.3](https://github.com/SeaQL/sea-orm/milestone/3) have been designated as "Transaction". This will be the focus of our effort in October.

**Comment by tyt2y3** [comment]
[0.3.0](https://github.com/SeaQL/sea-orm/releases/tag/0.3.0)
